Add new 3.2 features to aggregation builder

alcaeus opened the pull request:
This PR adds support for new aggregation features to the aggregation builder. You can read about these changes in the [MongoDB release notes](https://docs.mongodb.org/manual/release-notes/3.2/#aggregation-framework-enhancements).
To summarise:
* Added support for the $sample, $indexStats and $lookup stages
* Added support for new pipeline operators
* Added support for accumulators in the $project stage
* Add minDistance option to $geoNear stage
* Add new options to $unwind stage

ne0h12 commented:
Cool. I like it. When will ready new release with current features ?

alcaeus commented:
I'll add support for all other features added in 3.2 and then try to release a new version as soon as possible.

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Expr.php:
```diff
@@ -41,6 +42,23 @@ class Expr
+     * to a number.
+     *
+     * @see https://docs.mongodb.org/manual/reference/operator/aggregation/abs/
+     * @param mixed|self $number
```
Is "self" just here to suggest Expr? At first, it looked redundant alongside "mixed", but I can understand the reasoning.

alcaeus replied:
I mainly did that for all other classes as well, we might want to change that. Generally, you can pass:
- A string (denoting a field, like '$foo')
- An expression object (e.g. containing a $sum expression)
- An array (since you could've manually created an expression object)
- An actual number (int or float or MongoInt*) that will be passed to the pipeline.

Back when I wrote the methods I thought this would be the best idea to show what can be passed without specifying each type individually but still showing that you can pass an expression object as well.

jmikola replied:
:+1:

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Expr.php:
```diff
@@ -142,6 +160,26 @@ public function anyElementTrue($expression)
+     */
+    public function arrayElemAt($array, $index)
+    {
+        return $this->operator('$arrayElemAt', array($array, $index));
```
Did we bump 1.3 to PHP 5.5? If so, should this use short array syntax?

alcaeus replied:
Nice catch, will do that.

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Expr.php:
```diff
@@ -346,6 +437,25 @@ public function field($fieldName)
+     */
+    public function filter($input, $as, $cond)
+    {
+        return $this->operator('$filter', array('input' => $input, 'as' => $as, 'cond' => $cond));
```
Short array syntax?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Expr.php:
```diff
@@ -473,6 +616,62 @@ public function literal($value)
+     */
+    public function log($number, $base)
+    {
+        return $this->operator('$log', array($number, $base));
```
Short array syntax?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Expr.php:
```diff
@@ -473,6 +616,62 @@ public function literal($value)
+     */
+    public function log10($number)
+    {
+        return $this->operator('$log10', array($number));
```
Short array syntax?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Expr.php:
```diff
@@ -679,6 +878,26 @@ private function operator($operator, $expression)
+     */
+    public function pow($number, $exponent)
+    {
+        return $this->operator('$pow', array($number, $exponent));
```
Short array syntax? (I'll stop now)

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/IndexStats.php:
```diff
@@ -0,0 +1,41 @@
+    public function getExpression()
+    {
+        return array(
+            '$indexStats' => array()
```
Should this be `(object) []` or `new stdClass` to ensure it's serialized as a BSON document instead of an array?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Limit.php:
```diff
@@ -36,13 +37,13 @@ class Limit extends Stage
         parent::__construct($builder);
 
-        $this->limit = (integer) $limit;
+        $this->limit = (integer) $size;
```
Any reason we're renaming the argument but not the property?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Lookup.php:
```diff
@@ -0,0 +1,128 @@
+    /**
+     * @var string
+     */
+    protected $from;
```
Having looked at the other stage classes, should these be private by default?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Project.php:
```diff
@@ -60,4 +80,96 @@ public function includeFields(array $fields)
+     * @see https://docs.mongodb.org/manual/reference/operator/aggregation/stdDevPop/
+     * @see Expr::stdDevPop
+     * @param mixed|self $expression1
+     * @param mixed|self $expression2, ... Additional expressions
```
Should these be `mixed|Expr`? I don't think you'll be nesting Project stages.

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Group.php:
```diff
@@ -203,6 +204,44 @@ public function push($expression)
+     *
+     * @see https://docs.mongodb.org/manual/reference/operator/aggregation/stdDevPop/
+     * @see Expr::stdDevPop
+     * @param mixed|self $expression
```
`mixed|Expr`?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Group.php:
```diff
@@ -203,6 +204,44 @@ public function push($expression)
+     *
+     * @see https://docs.mongodb.org/manual/reference/operator/aggregation/stdDevSamp/
+     * @see Expr::stdDevSamp
+     * @param mixed|self $expression
```
`mixed|Expr`?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Operator.php:
```diff
@@ -46,6 +47,26 @@ public function __construct(Builder $builder)
+     *
+     * @see https://docs.mongodb.org/manual/reference/operator/aggregation/abs/
+     * @see Expr::abs
+     * @param mixed|self $number
```
`mixed|Expr`?

jmikola commented on lib/Doctrine/MongoDB/Aggregation/Stage/Project.php:
```diff
@@ -60,4 +80,96 @@ public function includeFields(array $fields)
+     * @see https://docs.mongodb.org/manual/reference/operator/aggregation/stdDevSamp/
+     * @see Expr::stdDevSamp
+     * @param mixed|self $expression1
+     * @param mixed|self $expression2, ... Additional expressions
```
`mixed|Expr`?

jmikola commented:
LGTM apart from the outstanding questions above.

alcaeus commented:
Fixed issues and rebased. Thanks @jmikola!

For the record, I only changed the most obvious instances to short array syntax. Once open PRs for the 1.3 release are merged I'll then go over the code and convert all remaining instances of `array(` to short syntax as well.

malarzm commented:
Hmmm @alcaeus feel free to remove my comments once resolved or good to be ignored, apparently when going commit by commit commenting works like this...

And nicely done I'll add! :+1:

alcaeus commented:
Merged manually in https://github.com/doctrine/mongodb/commit/4d3f8551b0603402196e6053499633dec18f2874.
